### ISSUE TITLE
ECS container ID not being parsed/sent

### ISSUE THREAD

**Issue by Nacoma** [elastic/apm-agent-php#823]
Infrastructure correlation isn't working because the `container.id` field is not present.

**To Reproduce**
- Add elastic-agent to the Elastic Beanstalk host
- Enable the docker module in metricbeat/elastic-agent
- Deploy a containerized PHP application to Elastic Beanstalk
- See
  - The container is incorrectly displayed as a host (duplicated, it's also displayed in containers). This container has no metrics.
  - Traces for the application have no associated infrastructure or metrics.

**Expected behavior**
- The container is not duplicated as a host
- Both host and container are associated with the trace

This is the ingested message received from the application/APM:

```json
{
  "@timestamp": [
    "2022-11-22T22:33:34.985Z"
  ],
  "agent.name": [
    "php"
  ],
  "agent.version": [
    "1.6.1"
  ],
  "data_stream.dataset": [
    "apm"
  ],
  "data_stream.namespace": [
    "develop"
  ],
  "data_stream.type": [
    "traces"
  ],
  "ecs.version": [
    "1.12.0"
  ],
  "event.agent_id_status": [
    "missing"
  ],
  "event.ingested": [
    "2022-11-22T22:33:36.000Z"
  ],
  "event.outcome": [
    "success"
  ],
  "host.hostname": [
    "fcf701cd93a8"
  ],
  "host.ip": [
    "172.17.0.2"
  ],
  "host.name": [
    "fcf701cd93a8"
  ],
  "http.request.method": [
    "POST"
  ],
  "observer.ephemeral_id": [
    "6ddef0a9-c6d4-4398-a08c-1c525d41e645"
  ],
  "observer.hostname": [
    "ip-blah.ec2.internal"
  ],
  "observer.id": [
    "77dcda3d-0f0e-4721-88fb-bad936adaaa3"
  ],
  "observer.type": [
    "apm-server"
  ],
  "observer.version": [
    "8.5.1"
  ],
  "process.pid": [
    53
  ],
  "processor.event": [
    "transaction"
  ],
  "processor.name": [
    "transaction"
  ],
  "service.environment": [
    "develop"
  ],
  "service.language.name": [
    "PHP"
  ],
  "service.language.version": [
    "8.1.12"
  ],
  "service.name": [
    "Api"
  ],
  "service.node.name": [
    "fcf701cd93a8"
  ],
  "service.runtime.name": [
    "PHP"
  ],
  "service.runtime.version": [
    "8.1.12"
  ],
  "timestamp.us": [
    1669156414985718
  ],
  "trace.id": [
    "57cfce17b3fc98f6f7405f60bd2ae45d"
  ],
  "transaction.duration.us": [
    111750
  ],
  "transaction.id": [
    "f473bb3a724dc27c"
  ],
  "transaction.name": [
    "POST /"
  ],
  "transaction.name.text": [
    "POST /"
  ],
  "transaction.result": [
    "HTTP 4xx"
  ],
  "transaction.sampled": [
    true
  ],
  "transaction.span_count.started": [
    0
  ],
  "transaction.type": [
    "request"
  ],
  "url.domain": [
  ],
  "url.full": [
  ],
  "url.full.text": [
  ],
  "url.original": [
  ],
  "url.original.text": [
  ],
  "url.path": [
    "/"
  ],
  "url.scheme": [
    "http"
  ],
  "_id": "Kwt5oYQBq_2iaxTGqjNm",
  "_index": ".ds-traces-apm-develop-2022.11.22-000001",
  "_score": null
}
```

Metricbeat (elastic-agent), installed on the host, is correctly identifying the containers:

```json
{
  "_index": ".ds-metrics-docker.container-develop-2022.11.22-000001",
  "_id": "_wuCoYQBq_2iaxTGSz52",
  "_version": 1,
  "_score": null,
  "fields": {
    "elastic_agent.version": [
      "8.5.1"
    ],
    "host.os.name.text": [
      "Amazon Linux"
    ],
    "host.hostname": [
      "ip-blah.ec2.internal"
    ],
    "host.mac": [
      "02-42-C2-AB-8F-BF",
      "02-6B-A7-10-33-CF",
      "42-7C-4E-97-2D-45",
      "CA-C4-80-19-42-B1"
    ],
    "docker.container.size.rw": [
      0
    ],
    "container.id": [
      "fcf701cd93a80cd5ae1ead16a80b871b0daea72d59868661b8bbef39d79dc1e6"
    ],
    "cloud.availability_zone": [
      "us-east-1a"
    ],
    "service.type": [
      "docker"
    ],
    "container.name": [
    ],
    "docker.container.labels.com_amazonaws_ecs_container-name": [
      "app"
    ],
    "container.image.name": [
    ],
    "host.os.version": [
      "2"
    ],
    "host.os.name": [
      "Amazon Linux"
    ],
    "agent.name": [
      "ip-blah.ec2.internal"
    ],
    "host.name": [
      "ip-blah.ec2.internal"
    ],
    "event.agent_id_status": [
      "verified"
    ],
    "docker.container.ip_addresses": [
      "172.17.0.2"
    ],
    "cloud.region": [
      "us-east-1"
    ],
    "host.os.type": [
      "linux"
    ],
    "data_stream.type": [
      "metrics"
    ],
    "docker.container.command": [
      "docker-php-entrypoint entrypoint"
    ],
    "host.architecture": [
      "x86_64"
    ],
    "cloud.machine.type": [
      "t3a.2xlarge"
    ],
    "cloud.provider": [
      "aws"
    ],
    "container.runtime": [
      "docker"
    ],
    "cloud.service.name": [
      "EC2"
    ],
    "agent.id": [
      "0c019f2c-a7a8-4d00-ab78-0d3bd41e4b93"
    ],
    "host.containerized": [
      false
    ],
    "ecs.version": [
      "8.0.0"
    ],
    "service.address": [
      "unix:///var/run/docker.sock"
    ],
    "docker.container.labels.com_amazonaws_ecs_task-arn": [
    ],
    "agent.version": [
      "8.5.1"
    ],
    "host.os.family": [
      "redhat"
    ],
    "docker.container.status": [
      "Up 34 minutes"
    ],
    "docker.container.labels.com_amazonaws_ecs_task-definition-version": [
      "410"
    ],
    "docker.container.created": [
      "2022-11-22T22:08:43.000Z"
    ],
    "docker.container.labels.com_amazonaws_ecs_task-definition-family": [
    ],
    "cloud.instance.id": [
      "i-0ca7c5a56b11067bf"
    ],
    "host.ip": [
      "10.0.1.253",
      "fe80::6b:a7ff:fe10:33cf",
      "172.17.0.1",
      "fe80::42:c2ff:feab:8fbf",
      "fe80::c8c4:80ff:fe19:42b1",
      "fe80::407c:4eff:fe97:2d45"
    ],
    "agent.type": [
      "metricbeat"
    ],
    "docker.container.labels.role": [
      "php"
    ],
    "event.module": [
      "docker"
    ],
    "host.os.kernel": [
      "4.14.275-207.503.amzn2.x86_64"
    ],
    "elastic_agent.snapshot": [
      false
    ],
    "host.id": [
      "9055fc78a53c4749b759e32a794d060f"
    ],
    "elastic_agent.id": [
      "0c019f2c-a7a8-4d00-ab78-0d3bd41e4b93"
    ],
    "data_stream.namespace": [
      "develop"
    ],
    "metricset.period": [
      10000
    ],
    "host.os.codename": [
      "Karoo"
    ],
    "event.duration": [
      2777816
    ],
    "metricset.name": [
      "container"
    ],
    "cloud.image.id": [
      "ami-03297dd2c45657fd8"
    ],
    "event.ingested": [
      "2022-11-22T22:43:01.000Z"
    ],
    "docker.container.size.root_fs": [
      0
    ],
    "@timestamp": [
      "2022-11-22T22:43:00.619Z"
    ],
    "host.os.platform": [
      "amzn"
    ],
    "cloud.account.id": [
      "121952031047"
    ],
    "data_stream.dataset": [
      "docker.container"
    ],
    "agent.ephemeral_id": [
      "b7ab80e1-fbe5-43b6-9bcc-ed55093f9d4b"
    ],
    "event.dataset": [
      "docker.container"
    ],
    "docker.container.labels.com_amazonaws_ecs_cluster": [
      "awseb-development-api-uimmpxrb6e"
    ]
  },
  "highlight": {
    "container.id": [
      "@kibana-highlighted-field@fcf701cd93a80cd5ae1ead16a80b871b0daea72d59868661b8bbef39d79dc1e6@/kibana-highlighted-field@"
    ]
  },
  "sort": [
    1669156980619
  ]
}
```

May be a different container ID now, but `/proc/self/cgroup` contains:

```
1:name=systemd:/ecs/9f72a38d047c4c4ab178664194e837c2/fcf701cd93a80cd5ae1ead16a80b871b0daea72d59868661b8bbef39d79dc1e6
```

**Comment by Nacoma** [comment]
Tangentially related, docker uses cgroup v2 (since docker 20.10) which does not expose the container ID in `/proc/self/cgroup`.

Instead, it looks like the container ID can be resolved from `/proc/self/mountinfo` in all versions. The same regex (`/\/docker\/containers\/(\w+)\//`) works in ECS as well.

**Comment by Nacoma** [comment]
Looks like this has come up in various elastic apm repos (https://github.com/elastic/apm-agent-python/issues/1691, https://github.com/elastic/apm/issues/523). https://github.com/elastic/apm/issues/523#issuecomment-943650430 is exactly the issue that we're running into.

**Comment by zobo** [comment]
Proposed PR #966

**Comment by SergeyKleyman** [comment]
Implemented by https://github.com/elastic/apm-agent-php/pull/966

**Comment by SergeyKleyman** [comment]
@Nacoma Hi. We implemented capturing container ID in PRs #966 and #975 and it's already merged to main branch. Please try it out and let us know if it works on a AWS ECS container. Also if you could provide examples of `/proc/self/cgroup` and `/proc/self/mountinfo` content from AWS ECS container that would allow us to add that dataset to the unit test to catch potential regressions in the future.

**Comment by Nacoma** [comment]
@SergeyKleyman 

> We implemented capturing container ID in PRs https://github.com/elastic/apm-agent-php/pull/966 and https://github.com/elastic/apm-agent-php/pull/975 and it's already merged to main branch.

That's wonderful! We have tickets in place to move to the new version. I appreciate the work you all are doing on this.

> Also if you could provide examples of /proc/self/cgroup and /proc/self/mountinfo content from AWS ECS container that would allow us to add that dataset to the unit test to catch potential regressions in the future.
Absolutely!

```
[ec2-user@ip-10-0-1-105 ~]$ cat /proc/self/cgroup
11:cpuset:/
10:pids:/user.slice
9:memory:/user.slice
8:cpu,cpuacct:/user.slice
7:blkio:/user.slice
6:hugetlb:/
5:net_cls,net_prio:/
4:perf_event:/
3:freezer:/
2:devices:/user.slice
1:name=systemd:/user.slice/user-1000.slice/session-9019.scope
```

```
[ec2-user@ip-10-0-1-105 ~]$ cat /proc/self/mountinfo
18 38 0:18 / /sys rw,nosuid,nodev,noexec,relatime shared:6 - sysfs sysfs rw
19 38 0:4 / /proc rw,nosuid,nodev,noexec,relatime shared:5 - proc proc rw
20 38 0:6 / /dev rw,nosuid shared:2 - devtmpfs devtmpfs rw,size=16324264k,nr_inodes=4081066,mode=755
21 18 0:7 / /sys/kernel/security rw,nosuid,nodev,noexec,relatime shared:7 - securityfs securityfs rw
22 20 0:19 / /dev/shm rw,nosuid,nodev shared:3 - tmpfs tmpfs rw
23 20 0:20 / /dev/pts rw,nosuid,noexec,relatime shared:4 - devpts devpts rw,gid=5,mode=620,ptmxmode=000
24 38 0:21 / /run rw,nosuid,nodev shared:21 - tmpfs tmpfs rw,mode=755
25 18 0:22 / /sys/fs/cgroup ro,nosuid,nodev,noexec shared:8 - tmpfs tmpfs ro,mode=755
26 25 0:23 / /sys/fs/cgroup/systemd rw,nosuid,nodev,noexec,relatime shared:9 - cgroup cgroup rw,xattr,release_agent=/usr/lib/systemd/systemd-cgroups-agent,name=systemd
27 18 0:24 / /sys/fs/pstore rw,nosuid,nodev,noexec,relatime shared:20 - pstore pstore rw
28 25 0:25 / /sys/fs/cgroup/devices rw,nosuid,nodev,noexec,relatime shared:10 - cgroup cgroup rw,devices
29 25 0:26 / /sys/fs/cgroup/freezer rw,nosuid,nodev,noexec,relatime shared:11 - cgroup cgroup rw,freezer
30 25 0:27 / /sys/fs/cgroup/perf_event rw,nosuid,nodev,noexec,relatime shared:12 - cgroup cgroup rw,perf_event
31 25 0:28 / /sys/fs/cgroup/net_cls,net_prio rw,nosuid,nodev,noexec,relatime shared:13 - cgroup cgroup rw,net_cls,net_prio
32 25 0:29 / /sys/fs/cgroup/hugetlb rw,nosuid,nodev,noexec,relatime shared:14 - cgroup cgroup rw,hugetlb
33 25 0:30 / /sys/fs/cgroup/blkio rw,nosuid,nodev,noexec,relatime shared:15 - cgroup cgroup rw,blkio
34 25 0:31 / /sys/fs/cgroup/cpu,cpuacct rw,nosuid,nodev,noexec,relatime shared:16 - cgroup cgroup rw,cpu,cpuacct
35 25 0:32 / /sys/fs/cgroup/memory rw,nosuid,nodev,noexec,relatime shared:17 - cgroup cgroup rw,memory
36 25 0:33 / /sys/fs/cgroup/pids rw,nosuid,nodev,noexec,relatime shared:18 - cgroup cgroup rw,pids
37 25 0:34 / /sys/fs/cgroup/cpuset rw,nosuid,nodev,noexec,relatime shared:19 - cgroup cgroup rw,cpuset
38 0 259:2 / / rw,noatime shared:1 - xfs /dev/nvme0n1p1 rw,attr2,inode64,noquota
17 19 0:17 / /proc/sys/fs/binfmt_misc rw,relatime shared:22 - autofs systemd-1 rw,fd=21,pgrp=1,timeout=0,minproto=5,maxproto=5,direct,pipe_ino=1457
40 20 0:16 / /dev/mqueue rw,relatime shared:23 - mqueue mqueue rw
39 20 0:35 / /dev/hugepages rw,relatime shared:24 - hugetlbfs hugetlbfs rw,pagesize=2M
41 18 0:9 / /sys/kernel/debug rw,relatime shared:25 - debugfs debugfs rw
43 38 0:37 / /var/lib/nfs/rpc_pipefs rw,relatime shared:26 - rpc_pipefs sunrpc rw
119 38 0:39 / /var/lib/docker/overlay2/fa814b48564c36ea763521c33dd30df2427d2da61d310d882f1969ac72c01045/merged rw,relatime shared:100 - overlay overlay rw,lowerdir=/var/lib/docker/overlay2/l/SRDYRBXB2QGUTE7NO3GMLYHADR:/var/lib/docker/overlay2/l/5KYEPXYAVGO6NWS4DBWDI6VCU4:/var/lib/docker/overlay2/l/7KE3LJBAEBS7JRMFEK5KF7BPQY:/var/lib/docker/overlay2/l/24GE4ZF5XD4LYR5X642743P362:/var/lib/docker/overlay2/l/QHRULB3RUD34SEMZZLY3NMOSFG:/var/lib/docker/overlay2/l/ES23L4VT3Z4VYHZL2VGW322KSH:/var/lib/docker/overlay2/l/KO256W4IZDKDUMW6I5BQIVWNFA,upperdir=/var/lib/docker/overlay2/fa814b48564c36ea763521c33dd30df2427d2da61d310d882f1969ac72c01045/diff,workdir=/var/lib/docker/overlay2/fa814b48564c36ea763521c33dd30df2427d2da61d310d882f1969ac72c01045/work
130 24 0:3 net:[4026531993] /run/docker/netns/default rw shared:103 - nsfs nsfs rw
133 38 0:40 / /var/lib/docker/containers/f6dc85b0f0f58ff346d32ddc9498f1855dd47450362d53b29c45cd25cf860e58/mounts/shm rw,nosuid,nodev,noexec,relatime shared:106 - tmpfs shm rw,size=65536k
79 38 0:38 / /var/lib/docker/overlay2/10e3787ccf8a6cf1f8465f309b41df61aed963468142e6b1a78e8bd6089eabfc/merged rw,relatime shared:61 - overlay overlay rw,lowerdir=/var/lib/docker/overlay2/l/LB63KJDCKVAGBBPXYROZX7JVAM:/var/lib/docker/overlay2/l/BVNPSN6Q6TYLNI2OHZWHTONQCY:/var/lib/docker/overlay2/l/X44I4X7YEIYQJQXO37U6P5DT6A:/var/lib/docker/overlay2/l/VJRZYCUF7DPR4R6RK4UL4SL6NG:/var/lib/docker/overlay2/l/2ODUSNSXD72XMD2FHZLCEIJZAG:/var/lib/docker/overlay2/l/YFYBKJUGA5QYPPNINOVVH2XOA4:/var/lib/docker/overlay2/l/U26ZYH2HQOKVGHCRY6O7N5II3V:/var/lib/docker/overlay2/l/JPHJFVCMI65D7ZNH4UBOKU7N6T:/var/lib/docker/overlay2/l/UVP75O73XHQCIDGZHQMJ43NSTG:/var/lib/docker/overlay2/l/JUUA3TYG6CC667ALFT75P4N36O:/var/lib/docker/overlay2/l/MMEXQLYH6QKEK4C3KWH25X27DM:/var/lib/docker/overlay2/l/4KMSKDQNAYQWCZRMGEFAINCVBT:/var/lib/docker/overlay2/l/Y3FK5XJATP4YDBBB2AYK3E5MK4:/var/lib/docker/overlay2/l/MC422ULYVLNVUIYQEQF4O7WSYQ:/var/lib/docker/overlay2/l/BHT7HF3GQSW2AHIIGZBWHAW422:/var/lib/docker/overlay2/l/HTL45WDTBKJJMJCCBY2VF2DTPP:/var/lib/docker/overlay2/l/QQ7Y732UMNN2H3CLTT6RPRI7YG:/var/lib/docker/overlay2/l/AOZL7ABTR7JHFHXPOLZTOHAVJ2:/var/lib/docker/overlay2/l/IACECLNNMDX35D3LJDHYKVTTGC:/var/lib/docker/overlay2/l/XIWRLIGDY3T7NGIYL7TP4FFRDM:/var/lib/docker/overlay2/l/B4OQCGLKMLA43TIJ7GSZHYRCZJ:/var/lib/docker/overlay2/l/RLMQKM54FCVWHKOIFNO3RQXPP6:/var/lib/docker/overlay2/l/OR5WBD6GPVBQQOMGVBA6ZDTXIL:/var/lib/docker/overlay2/l/XFEDB4UMUYTJEV53DYEJQRAG2S:/var/lib/docker/overlay2/l/7PHSVW2LCKORUIQBRG6DH27XDM:/var/lib/docker/overlay2/l/C35VMUOBUUSXGMH2QJQOYTWFT6:/var/lib/docker/overlay2/l/6JSGLPR65IN2FNCF2LLXALB2F2:/var/lib/docker/overlay2/l/2RWWAASWNVH7R3WXT3ZDDNQJGR,upperdir=/var/lib/docker/overlay2/10e3787ccf8a6cf1f8465f309b41df61aed963468142e6b1a78e8bd6089eabfc/diff,workdir=/var/lib/docker/overlay2/10e3787ccf8a6cf1f8465f309b41df61aed963468142e6b1a78e8bd6089eabfc/work
110 38 0:47 / /var/lib/docker/containers/837c1bc37405aee683ad65cbc4fe95f5e5e918ac73ae84afb26ddff2415b9367/mounts/shm rw,nosuid,nodev,noexec,relatime shared:63 - tmpfs shm rw,size=65536k
247 24 0:3 net:[4026532258] /run/docker/netns/0984df741f1c rw shared:65 - nsfs nsfs rw
112 38 0:56 / /var/lib/docker/overlay2/582437659e7b8a17f1eaccc9ec47f7a3b42d34de2748ff7e45c16b632dae2dfa/merged rw,relatime shared:67 - overlay overlay rw,lowerdir=/var/lib/docker/overlay2/l/TYRODBSIYP3HMTHBHD7WZHPQV2:/var/lib/docker/overlay2/l/VGP7JNR7LBNRX5YYQ6ZPE2TT6D:/var/lib/docker/overlay2/l/5GDA6GNDQJC6CCQ4DSERQBGOWV:/var/lib/docker/overlay2/l/XQSYVAHZL5DKE3B4FBXDNHYMBL:/var/lib/docker/overlay2/l/YINLL3Y4VDVC4YX6GCY64KDDOM:/var/lib/docker/overlay2/l/NPRO4YEZAXRVDHEX42GVTXXLDM:/var/lib/docker/overlay2/l/GG6OO7MRJNWABIZ6H3WEGVK56R,upperdir=/var/lib/docker/overlay2/582437659e7b8a17f1eaccc9ec47f7a3b42d34de2748ff7e45c16b632dae2dfa/diff,workdir=/var/lib/docker/overlay2/582437659e7b8a17f1eaccc9ec47f7a3b42d34de2748ff7e45c16b632dae2dfa/work
162 38 0:57 / /var/lib/docker/containers/42f87e7a54798fbc9bc1af49a2ae9c9168d5aab2671a9ff31eedf5d982843ac8/mounts/shm rw,nosuid,nodev,noexec,relatime shared:69 - tmpfs shm rw,size=65536k
307 24 0:3 net:[4026532341] /run/docker/netns/8c081e396c48 rw shared:71 - nsfs nsfs rw
164 24 0:66 / /run/user/1000 rw,nosuid,nodev,relatime shared:73 - tmpfs tmpfs rw,size=3266420k,mode=700,uid=1000,gid=1000
```
